core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1090,38 +1090,29 @@ class ReplicaManager(val config: KafkaConfig,
    * @param producerId      the producer id for the producer writing to the transaction
    * @param producerEpoch   the epoch of the producer writing to the transaction
    * @param baseSequence    the base sequence of the first record in the batch we are trying to append
-   * @param requestLocal    container for the stateful instances scoped to this request -- this must correspond to the
-   *                        thread calling this method
    * @param callback        the method to execute once the verification is either completed or returns an error
    *
    * When the verification returns, the callback will be supplied the error if it exists or Errors.NONE.
    * If the verification guard exists, it will also be supplied. Otherwise the SENTINEL verification guard will be returned.
-   * This guard can not be used for verification and any appends that attenpt to use it will fail.
+   * This guard can not be used for verification and any appends that attempt to use it will fail.
    */
   def maybeStartTransactionVerificationForPartition(
     topicPartition: TopicPartition,
     transactionalId: String,
     producerId: Long,
     producerEpoch: Short,
     baseSequence: Int,
-    requestLocal: RequestLocal,
-    callback: (Errors, RequestLocal, VerificationGuard) => Unit
+    callback: Either[Errors, VerificationGuard] => Unit
   ): Unit = {
-    def generalizedCallback(preAppendErrors: Map[TopicPartition, Errors],
-                            newRequestLocal: RequestLocal,
-                            verificationGuards: Map[TopicPartition, VerificationGuard]): Unit = {
-      callback(
-        preAppendErrors.getOrElse(topicPartition, Errors.NONE),
-        newRequestLocal,
-        verificationGuards.getOrElse(topicPartition, VerificationGuard.SENTINEL))
+    def generalizedCallback(results: Map[TopicPartition, Either[Errors, VerificationGuard]]): Unit = {
```

Comment:
I think we could do the translation from preAppendErrors, newRequestLocal, verificationGuards here, then we'd avoid propagating the changes all the way to replication layer.

Reply:
Yeah, I was debating about this. The reasoning for not doing it here is that @jolshan may also use `maybeStartTransactionVerificationForPartitions` from the core path in conjunction with `wrapAsyncCallback` so she will also need to have a unary callback. @jolshan Is my understanding correct? Otherwise, we could limit the change to this method.

Reply:
Please take a look at my refactor PR. I have some this to some extent.
I'd prefer not to overhaul it again (as I did after the previous group coordinator change)
Hopefully it makes this work easier too.

Reply:
https://github.com/apache/kafka/pull/15087

Reply:
Ack. I will take a look at https://github.com/apache/kafka/pull/15087.